Broadcast Receivers Tests and PowerMock

Cogno-Marco opened the pull request:
## Premise
During recent discussions in other pull request/issues (I think in #22 for example) one thing was strange: 2 methods used in SMSCore used only for unit-tests.
In this pull-request I've removed those methods, and I've imported PowerMock to use instead of those 2 methods to test the static `SmsManager.getDefault()` method.
At the time of the creation of those 2 methods I wasn't yet aware of the existance of PowerMock, and when I discovered it I was not able to make it work. This has recently changed, since I was finally able to correctly use it.

I've also fully tested (I think) `SMSSendBroadcastReceiver` and `SMSDeliveredBroadcastReceiver`, but I was not able to test `SMSReceivedBroadcastReceiver` due to the relative complexity using mocks, since that receiver is motly calling android APIs

Thanks to these changes both of those receivers have reached 100% code coverage

## Changes

### 1. Imported PowerMock
For PowerMock to work I needed to correctly used a version of PowerMockito compatible with the version used of Mockito.

### 2. Removed methods in SMSCore
Since there's now PowerMock, those methods in SMSCore are redundant (you can use PowerMock instead), so I've removed them.
### 3. Fixed SMSManager unit-tests
Using PowerMock I fixed broken unit-tests since I've removed those functions in SMSCore
### 4. Tested `SMSSendBroadcastReceiver` and `SMSDeliveredBroadcastReceiver`
I've tests those 2 classes using Mock objects

## Side changes
### AndroidPreferences module
I removed a useless AndroidPreference module import, since a couple of rows before that there's already one present and some versions updated.
___
With the next one I was able to write "PowerMock" 10 times 🎉

giovannivelludo committed:
added files to .gitignore

giovannivelludo committed:
removed modules.xml, as it was added to .gitignore

giovannivelludo committed:
added to .gitignore all patterns present in default .gitignore generated by Android Studio

giovannivelludo committed:
Merge branch 'hotfix/gitignore'

giovannivelludo committed:
Merge pull request #1 from Cogno-IDU/master

updated .gitignore (#8)

Cogno-Marco committed:
Merge remote-tracking branch 'central/master'

giovannivelludo committed:
Merge remote-tracking branch 'upstream/master'

Cogno-Marco committed:
Merge remote-tracking branch 'central/master'

Cogno-Marco committed:
Merge remote-tracking branch 'central/master'

matteocarnelos committed:
Merge remote-tracking branch 'central/master'

Cogno-Marco committed:
Merge remote-tracking branch 'central/master'

Cogno-Marco committed:
Merge remote-tracking branch 'central/master'

Cogno-Marco committed:
Imported PowerMock
Removed useless core functions which were used only in unit-tests
Used PowerMock to still test functionality whitout using those functions
Fully Tested SMSSendBroadcastReceiver

Cogno-Marco committed:
Fully Tested SMSDeliveredBroadcastReceiver

Cogno-Marco committed:
Removed Unused imports

Cogno-Marco committed:
Removed useless local variable

giovannivelludo commented:
Changes 1 and 2 are duplicates of https://github.com/Cogno-IDU/smslibrary/pull/20

AlbertoUrsino commented on smslibrary/src/test/java/com/eis/smslibrary/SMSDeliveredBroadcastReceiverTest.java:
```diff
@@ -0,0 +1,65 @@
+        receiver.onReceive(contextMock, null);
+        verify(listenerMock, times(1))
+                .onSMSDelivered(VALID_MESSAGE, SMSMessage.DeliveredState.MESSAGE_DELIVERED);
+        verify(contextMock).unregisterReceiver(receiver);
```
Well, now I try to write what you did to understand well, tell me if I'm wrong:
1) Force the `getResultCode ()` method, once called, to return `Activity.RESULT_OK`
2) Invoke the `onReceive` method which should assign the` MESSAGE_DELIVERED` value to the `DeliveryState` variable
3) Check that the `onSMSDelivered` method has been called
4) Check that the `unregisterReceiver` method has been called

I don't quite understand what the `times (x)` parameter is for...

Cogno-IDU replied:
yup, you're definitely right. When `getResultCode()` is called it will automatically return `Activity.RESULT_OK`
(**N.B.**, I want the `BroadcastReceiver` to do so, not the `SMSDelibveredBroadcastReceiver`, so I need that cast.
then I'll call the function and verify is eveything went according to plans

The `times(1)` is used to check how many times the function was called, so `times(5)` will check if the function was called 5 times

Cogno-IDU replied:
Another insteresting thing. `SMSSentBroadcastReceiver` (and Delivered too) are **NOT** mocks, they are Spies, the difference is that a mock is an object which when a non-mocked function is called nothing is returned or done, while a spy is an object which will work as a normal oobject but where I can **replace** functions with mocks

AlbertoUrsino commented on smslibrary/src/test/java/com/eis/smslibrary/SMSSentBroadcastReceiverTest.java:
```diff
@@ -0,0 +1,97 @@
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
```
Author :)

AlbertoUrsino commented on smslibrary/src/test/java/com/eis/smslibrary/SMSSentBroadcastReceiverTest.java:
```diff
@@ -0,0 +1,97 @@
+import static org.mockito.Mockito.verify;
+
+public class SMSSentBroadcastReceiverTest {
+
```
In the first test class you didn't insert this space

albertoursino reviewed (changes requested):
Thanks to you @Cogno-IDU I understood a little bit how PowerMocks work.

AlbertoUrsino commented on smslibrary/src/test/java/com/eis/smslibrary/SMSDeliveredBroadcastReceiverTest.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Test
+    public void constructor_doesntThrow(){
+        new SMSDeliveredBroadcastReceiver(VALID_MESSAGE, listenerMock);
```
The "automatic code reviewer" is in right.
Here should write:
`try {`
     `new SMSDeliveredBroadcastReceiver(VALID_MESSAGE, listenerMock);`
     `//Success }`
`catch (Exception e) `{
     `Assert.fail(...); }`

Cogno-IDU replied:
Well, if the construction returns an error, the test will automatically fail, so I wouldn't need to check that, maybe there's a better way?

Cogno-IDU replied:
I checked online and couldn't find any smarter way to do it, so I'll use this so that codacy won't complain

AlbertoUrsino commented on smslibrary/src/test/java/com/eis/smslibrary/SMSSentBroadcastReceiverTest.java:
```diff
@@ -0,0 +1,97 @@
+
+    @Test
+    public void constructor_doesntThrow(){
+        new SMSSentBroadcastReceiver(VALID_MESSAGE, listenerMock);
```
The "automatic code reviewer" is in right.
Here should write:
`try {`
     `new SMSSentBroadcastReceiver(VALID_MESSAGE, listenerMock);`
     `//Success }`
`catch (Exception e) `{
     `Assert.fail(...); }`

albertoursino reviewed (changes requested):
2nd part of the review

giovannivelludo commented:
Duplicate of #20

Cogno-Marco committed:
rewritten constructor test due to codacy error
formatted both classes
added author tags
renamed SMSManager from instance to smsManager

giovannivelludo commented:
Duplicate of #20

albertoursino reviewed (approved):
Well done!

Cogno-IDU commented:
> Duplicate of #20

Not a duplicate, only the first 2 files are (basically integration of PowerMock and removing SMSCore Methods)
We'll keep both pr opened and the first to be integrated will have those 2 files, then the one to still be open will need to update it with changes added

giovannivelludo commented:
@Cogno-IDU ok, could you please add me as author in `SMSCore` and `SMSManagerTest`?

giovannivelludo commented on smslibrary/build.gradle:
```diff
@@ -37,5 +37,10 @@ dependencies {
-    implementation 'com.github.CremaLuca:android-preferences:2.3'
+
+    androidTestImplementation 'com.android.support.test:rules:1.0.2'
+    androidTestImplementation 'org.mockito:mockito-android:3.2.4'
```
These two dependencies are not needed, as you can see they're for instrumented tests but we don't have any of those. I had added them temporarily in #20 when I was writing instrumented test, but I couldn't get them to compile so I removed them in the end.

giovannivelludo commented on smslibrary/src/test/java/com/eis/smslibrary/SMSManagerTest.java:
```diff
@@ -25,7 +30,13 @@
+    public void setup(){
+        PowerMockito.mockStatic(SmsManager.class);
+        when(SmsManager.getDefault()).thenReturn(managerMock);
+    }
```
I'll copy this in #20 , `nullMessage_throws()` was working even without PowerMockito, but now I see that the `NullPointerException` was being thrown for the wrong reason.

giovannivelludo commented on smslibrary/src/test/java/com/eis/smslibrary/SMSSentBroadcastReceiverTest.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Test
+    public void receive_genericFailure_correctListener() {
+        doReturn(666)
```
:(

Cogno-IDU replied:
¯\_(ツ)_/¯

giovannivelludo replied:
You can mark this as resolved

Cogno-Marco committed:
- [gradle] Removed useless imports
+ [SMSCore/SMSManagerTest] added author tags
+ [Sent/Delivered BR] improved documentation

giovannivelludo commented on smslibrary/src/test/java/com/eis/smslibrary/SMSManagerTest.java:
```diff
@@ -34,13 +45,12 @@ public void singletonInstance() {
     @Test(expected = NullPointerException.class)
     public void nullMessage_throws() {
-        instance.sendMessage(null);
+        smsManager.sendMessage(null);
```
Wait a second, I was wrong in https://github.com/Cogno-IDU/smslibrary/pull/39#discussion_r362356126, the `NullPointerException` is thrown in `DefaultSMSMessageParseStrategy.parseData()`, because it calls `message.getData()` and `message` points to `null`, so method `setup()` in this test doesn't need to be executed before every test, you can move its body to method `validMessage_isSent()`, since it's the only one using mocks.

Cogno-IDU replied:
Since it's already working I don't think this kind of edit is necessary. Also in general is not a particularly good practice, what if I add another test which needs setup() ? it obviously won't work unless I call it, but then I'll need to re-create `setup()`. So I think I'll leave it there

giovannivelludo replied:
Ok, I thought methods annotated with `@Before` had to contain only setup operations that need to be performed before every test in the class. In cases where a lot of tests need to be run and performance matters, it's probably better to use `@BeforeClass` for operations that need to be executed only once, before all tests are executed.

Cogno-Marco committed:
Merge remote-tracking branch 'central/master'

Cogno-Marco committed:
Merge branch 'master' into feature/powermock-tests

# Conflicts:
#	smslibrary/build.gradle
#	smslibrary/src/main/java/com/eis/smslibrary/SMSCore.java
#	smslibrary/src/test/java/com/eis/smslibrary/SMSManagerTest.java

Cogno-IDU commented:
I've updated the branch due to recent changes, but now I have to fix and re-work some stuff due to those changes.

Cogno-Marco committed:
Fixed unit-tests:
~ Fixed unit tests in SMSSentBroadcastReceiver and SMSDeliveredBroadcastReceiver

Cogno-IDU commented:
While I was fixing these I ran all the tests and I noticed an error in a test
@giovannivelludo The test `validMessage_isSent()` in `SMSManagerTest` is not working because due to your #20 pr, you changes how messages are sent (from messages to multipart messages) so the test is failing.
1. I think it should have been an error fixed in you pr, but other than that (this is not really a problem imo, it can happen to everyone to miss an error)
2. Didn't the pross say to not use multipart sms messages? why are we using those then?

Anyway, I didn't fix the error here, so it needs to be fixed somewhere else.

giovannivelludo commented:
All tests were working when I submitted the last commit in #20 , otherwise Codacy would have pointed that out. I ran them again right now and they still work. What exactly did I mess up?
![Screenshot from 2020-01-06 19-35-00](https://user-images.githubusercontent.com/56405609/71839552-b3b19a00-30bb-11ea-8bb0-58426435bca1.png)


Professor Peserico said we shouldn't use multipart messages because some carriers charge more for those than regular messages, but I tested them on phones with Vodafone, Tre, Wind, Iliad and Coop Voce SIM cards and nobody was charged more than for regular messages. All contracts included unlimited or a certain number of messages, and multi-part messages were counted correctly. If anybody finds an offer with unlimited messages where multiparts have additional charges compared to regular text messages then I agree with not using them. Besides, the library doesn't force anybody to use multi part messages, messages that fit in a single SMS will be sent as a single SMS.

Another argument brought up by De Zen was that when he had a contract with Tre, whenever he wrote a message longer than 3 SMS it was automatically converted to an MMS. However [the conversion was likely made by the app he was using](https://android.stackexchange.com/questions/4142/how-can-i-prevent-my-device-from-automatically-converting-sms-to-mms) to send messages.

giovannivelludo commented on smslibrary/src/test/java/com/eis/smslibrary/SMSManagerTest.java:
```diff
@@ -53,50 +45,12 @@ public void singletonInstance() {
-                isNull(String.class), messageTextCaptor.capture(), isNull(ArrayList.class),
-                isNull(ArrayList.class));
-        assertEquals(peerNumberCaptor.getValue(), VALID_PEER_NUMBER);
-        assertEquals(messageTextCaptor.getValue(), messageText);
```
Why did you remove all these lines of tests? I think the reason you have non-working tests is not because of #20 (they were all working there), but because something went wrong when you merged [`master` from the main repository](/Cogno-IDU/smslibrary/tree/master) into [your branch](/EIS0/smslibrary/tree/feature/powermock-tests).

Cogno-IDU commented:
Oh ok, I think I understand what happened, I'll fix it when I can over the next few days 👍

Cogno-Marco committed:
fix: Re-enstablished tests lost during a merge

Cogno-IDU commented:
@giovannivelludo I finally fixed this, now everything is working correctly 👍

giovannivelludo commented on smslibrary/src/test/java/com/eis/smslibrary/SMSManagerTest.java:
```diff
@@ -32,20 +33,29 @@
     private final SMSPeer VALID_PEER = new SMSPeer(VALID_PEER_NUMBER);
     private final SMSMessage VALID_MESSAGE = new SMSMessage(VALID_PEER, VALID_MESSAGE_TEXT);
-    private final SMSManager managerInstance = SMSManager.getInstance();
+    private SMSManager managerInstance = SMSManager.getInstance();
```
why is it not final anymore?

giovannivelludo commented on smslibrary/src/test/java/com/eis/smslibrary/SMSManagerTest.java:
```diff
@@ -32,20 +33,29 @@
+
     private final String VALID_PEER_NUMBER = "+393423541601";
     private final String VALID_MESSAGE_TEXT = "valid text";
+    private final String SENT_TEXT = (char) 0x02 + VALID_MESSAGE_TEXT;
```
this variable is not used

giovannivelludo commented on smslibrary/src/test/java/com/eis/smslibrary/SMSManagerTest.java:
```diff
@@ -32,20 +33,29 @@
     @Captor
     private ArgumentCaptor<String> peerNumberCaptor;
 
+    private SmsManager managerMock = mock(SmsManager.class);
```
can be final

giovannivelludo commented on smslibrary/src/test/java/com/eis/smslibrary/SMSManagerTest.java:
```diff
@@ -17,8 +18,8 @@
 import static org.mockito.Matchers.isNull;
+import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
-import static org.powermock.api.mockito.PowerMockito.mock;
```
what's the reason for this change?

Cogno-IDU replied:
Mock objects should come from Mockito instead of PowerMockito, that should be used only for static functions.

giovannivelludo replied:
`org.powermock.api.mockito.PowerMockito.mock()` mocks classes just like Mockito's `mock()`, not static methods. `org.powermock.api.mockito.PowerMockito.mockStatic()` is the method used for mocking a class' static methods.
Anyway, I just read the documentation for both `mock()` methods and PowerMockito's version supports mocking of final and native methods, which I guess is not possible with Mockito since this is not mentioned there. So I would keep Mockito's version because it's more restrictive, you can mark this as resolved.

Cogno-Marco committed:
style: made variables final
fix: removed unused variable

Cogno-IDU commented on smslibrary/src/test/java/com/eis/smslibrary/SMSManagerTest.java:
```diff
@@ -36,14 +36,13 @@
     private final SMSPeer VALID_PEER = new SMSPeer(VALID_PEER_NUMBER);
     private final SMSMessage VALID_MESSAGE = new SMSMessage(VALID_PEER, VALID_MESSAGE_TEXT);
-    private SMSManager managerInstance = SMSManager.getInstance();
+    private final SMSManager MANAGER_INSTANCE = SMSManager.getInstance();
```
`final` marks constants
also if a variable is marked as `final`, every object instance of the class sees that variable in the same way, effectively making it static

giovannivelludo replied:
>final marks constants

This is not true, from the official Java Coding Style Guide I linked:
> The
following are considered to be constants:
>1. All static final primitive types (Remember that all interface fields are inherently static
final).
>2. All static final object reference types that are never followed by "." (dot).
>3. All static final arrays that are never followed by "[" (dot).

As you can see, final fields that are not also static are not considered constants and should not be all upper-case.

RicDeZen replied:
I don't feel like arguing over it but Giovanni has a point. I personally only put in full capital both `static` and `final` fields.

Cogno-IDU replied:
The only difference between a final variable and a static final variable is this:
```
public class Foo{
    public final int a = 5;
    public static final int b = 10;
}

//outside method
public void barMethod(){
    Foo fooInstance = new Foo();
    fooInstance.a; // returns 5
    Foo.b; //returns 10
    fooInstance.a = 20; //ERROR
    Foo.b = 30; //ERROR
}
```

So yeah, the only difference is that you can access the `static final` variable without needing to instantiate the class, while you need that instance to get that `final` variable.

But in our case that variable is private, so it can't be accessed from outside, making this difference irrelevant, so effectively making it a constant

giovannivelludo replied:
> But in our case that variable is private, so it can't be accessed from outside, making this difference irrelevant, so effectively making it a constant

No, the definition of "constant" does not contemplate object references without the `static` keyword. Even if `MANAGER_INSTANCE` is initialized when the class is loaded, just like a constant, it cannot be a constant because of its definition.

Cogno-IDU replied:
Yeah ok, I get it, you only want to follow the definition, without thinking abuot the problem itself, alright, I'll make the change, there's no point in arguing for this.
But to me that's still a constant:
[If it looks like a duck, swims like a duck, and quacks like a duck, then it probably is a duck.](https://en.wikipedia.org/wiki/Duck_test)

giovannivelludo replied:
Even when you look at how a `private final` field and a `private static final` variable are treated by the JVM they're different, a `static` variable is initialized only once at the start of execution by the ClassLoader, while a `private final` field is initialized every time an instance of the object is created.

ricdezen reviewed (approved):
I forgot this pull was still open. I was planning on doing these myself but these are no different from what I'd do.

Cogno-Marco committed:
feat: changed constants from final to static final

giovannivelludo commented on smslibrary/src/test/java/com/eis/smslibrary/SMSManagerTest.java:
```diff
@@ -62,7 +71,7 @@ public void validMessage_isSent() {
         when(SmsManager.getDefault()).thenReturn(managerMock);
 
-        managerInstance.sendMessage(VALID_MESSAGE);
+        MANAGER_INSTANCE.sendMessage(VALID_MESSAGE);
```
This is not a [constant](https://web.archive.org/web/20120911192801if_/http://developers.sun.com/sunstudio/products/archive/whitepapers/java-style.pdf#G3.5938) and should not be all upper-case:
> The
following are considered to be constants:
>1. All static final primitive types (Remember that all interface fields are inherently static
final).
>2. **All static final object reference types that are never followed by "." (dot)**.
>3. All static final arrays that are never followed by "[" (dot).

Cogno-IDU replied:
What do you mean?
`private static final SMSManager MANAGER_INSTANCE = SMSManager.getInstance();`

giovannivelludo replied:
In the line I mentioned, `MANAGER_INSTANCE` is followed by a dot. Therefore it's not an object reference type that is never followed by a dot, therefore it's not a constant.

Cogno-IDU replied:
Where did you even find that definition?

Cogno-IDU replied:
[Oh btw](https://docs.oracle.com/javase/specs/jls/se13/html/jls-6.html)

> Constant Names
> The names of constants in interface types should be, and final variables of class types may conventionally be, a sequence of one or more words, acronyms, or abbreviations, all uppercase, with components separated by underscore "_" characters. Constant names should be descriptive and not unnecessarily abbreviated. Conventionally they may be any appropriate part of speech.

>  final variables of class types may conventionally be a sequence of one or more words[...] all uppercase, with components separated by underscore "_" characters.

giovannivelludo replied:
Sun's coding style guide is from the year 2000, while the Oracle's language specification you found is a lot more recent, so I'm okay with going with that

giovannivelludo commented on smslibrary/src/test/java/com/eis/smslibrary/SMSSentBroadcastReceiverTest.java:
```diff
@@ -0,0 +1,110 @@
+    private final SMSPeer VALID_PEER = new SMSPeer("+393423541601");
+    private final String VALID_TEXT = "Some random text to send";
+    private final ArrayList<String> MESSAGE_TO_SEND = new ArrayList<>();
+    private final SMSMessage VALID_MESSAGE = new SMSMessage(VALID_PEER, VALID_TEXT);
```
These are not constants and should not be all upper-case

Cogno-Marco committed:
style: changed final variables to static final
fix: fixed consecutive tests throwing errors due to bad setups

Cogno-Marco committed:
style: changed some variables naming style

francescobau commented on smslibrary/src/test/java/com/eis/smslibrary/SMSDeliveredBroadcastReceiverTest.java:
```diff
@@ -0,0 +1,79 @@
+
+    @Test
+    public void constructor_doesntThrow() {
+        try {
+            new SMSDeliveredBroadcastReceiver(MESSAGE_TO_SEND, listenerMock, VALID_PEER);
+            //Success
+        } catch (Exception e) {
+            fail();
+        }
+    }
```
Test fails automatically when an Exception is thrown.
Is this `try-catch` necessary?
It would be appreciated if you let me know...

Cogno-Marco replied:
the try-catch is not _necessarily_ required, but it's used to clearly know what I'm testing for

francescobau commented on smslibrary/src/test/java/com/eis/smslibrary/SMSSentBroadcastReceiverTest.java:
```diff
@@ -0,0 +1,111 @@
+
+    @Test
+    public void constructor_doesntThrow() {
+        try {
+            new SMSSentBroadcastReceiver(MESSAGE_TO_SEND, listenerMock, VALID_PEER);
+            //Success
+        } catch (Exception e) {
+            fail();
+        }
```
Same thing as above.

francescobau reviewed (commented):
@Cogno-IDU 
Hi there.
Sorry for the late review.
I was pleased to see how you use PowerMock, but I've got just 2 doubts in here... one of them is expressed in the comments, but the other one is `verify(mockVariable, times(1))` ... I can see that number `1` literally everywhere, and for the same purpose. If that's supposed to be only one every time, wouldn't it be better if you made it a constant?
Can that be considered a hardwiring issue?
Thank you in advance.

P.S. I know, it's just a test class, but I'm just asking ;)
